src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/AbstractLanguageServer.cs

```diff
@@ -0,0 +1,327 @@
+    }
+
+    public virtual async Task ExitAsync()
+    {
+        try
+        {
+            await ShutdownRequestExecutionQueueAsync().ConfigureAwait(false);
+
+            var lspServices = GetLspServices();
+            lspServices.Dispose();
+
+            _jsonRpc.Disconnected -= JsonRpc_Disconnected;
+            _jsonRpc.Dispose();
+        }
+        catch (Exception)
+        {
+            // Swallow exceptions thrown by disposing our JsonRpc object. Disconnected events can potentially throw their own exceptions so
+            // we purposefully ignore all of those exceptions in an effort to shutdown gracefully.
+        }
+    }
+
+    private ValueTask ShutdownRequestExecutionQueueAsync()
+    {
+        var queue = GetRequestExecutionQueue();
+        return queue.DisposeAsync();
+    }
+
+    protected virtual void RequestExecutionQueueErroredInternal(string message)
+    {
+    }
+
+#pragma warning disable VSTHRD100
+    private async void RequestExecutionQueue_Errored(object? sender, RequestShutdownEventArgs e)
+    {
+        // log message and shut down
+        await _logger.LogWarningAsync($"Request queue is requesting shutdown due to error: {e.Message}", CancellationToken.None).ConfigureAwait(false);
+
+        RequestExecutionQueueErroredInternal(e.Message);
+
+        var lspServices = GetLspServices();
+
+        // We want to sue the LifecycleManager to ensure we fire the events
+        var lifeCycleManager = lspServices.GetRequiredService<LifeCycleManager<RequestContextType>>();
+        await lifeCycleManager.ShutdownAsync(e.Message).ConfigureAwait(false);
+        await lifeCycleManager.ExitAsync().ConfigureAwait(false);
+    }
+
+    /// <summary>
+    /// Cleanup the server if we encounter a json rpc disconnect so that we can be restarted later.
+    /// </summary>
+    private async void JsonRpc_Disconnected(object? sender, JsonRpcDisconnectedEventArgs e)
+    {
+        if (_shuttingDown)
+        {
+            // We're already in the normal shutdown -> exit path, no need to do anything.
+            return;
+        }
+
+        var message = $"Encountered unexpected jsonrpc disconnect, Reason={e.Reason}, Description={e.Description}, Exception={e.Exception}";
+        await _logger.LogWarningAsync(message, CancellationToken.None).ConfigureAwait(false);
+
+        var lspServices = GetLspServices();
+        var lifeCycleManager = lspServices.GetRequiredService<LifeCycleManager<RequestContextType>>();
+
+        await lifeCycleManager.ShutdownAsync(message).ConfigureAwait(false);
+        await lifeCycleManager.ExitAsync().ConfigureAwait(false);
+    }
+#pragma warning disable VSTHRD100
+
+    /// <summary>
+    /// Disposes the LanguageServer, clearing and shutting down the queue and exiting.
+    /// Can be called if the Server needs to be shut down outside of 'shutdown' and 'exit' requests.
+    /// </summary>
+    public virtual async ValueTask DisposeAsync()
```

Comment:
would not make this virtual and instead leave customization up to the life cycle manager (if possible)